### ISSUE TITLE
TypeScript errors with --strictFunctionTypes flag.

### ISSUE THREAD

**Issue by whhhooop** [reduxjs/reselect#391]
reselect: 4.0.0
typescript: 3.2.2

The following code produces TS error.
```
// tsconfig.json
{
  "compilerOptions": {
    "strictFunctionTypes": true
  }
}
```
```
// app.ts
import { createStructuredSelector } from 'reselect';

type GlobalState = {
  foo: string;
  bar: number;
};

const selectFoo = (state: GlobalState) => state.foo;
const selectBar = (state: GlobalState) => state.bar;

const injectState = createStructuredSelector({
  foo: selectFoo, // error
  bar: selectBar, // error
});

```
Error text
```
Type '(state: GlobalState) => string' is not assignable to type 'ParametricSelector<{}, {}, string>'.
  Types of parameters 'state' and 'state' are incompatible.
    Type '{}' is missing the following properties from type 'GlobalState': foo, bar [2322]
```
With `strictFunctionTypes` disabled, everything works fine. What is the correct way of using `createStructuredSelector` in TypeScript with`strictFunctionTypes` option?

**Comment by OliverJAsh** [comment]
I also just ran into this. Filed a TypeScript issue. See the issue for details on workaround: https://github.com/Microsoft/TypeScript/issues/29479.

**Comment by markerikson** [comment]
Is this still an issue with TS 4.x?

**Comment by whhhooop** [comment]
Yes. Just tested this example with TS 4.1.2 and reselect 4.0.0

**Comment by markerikson** [comment]
Okay.  Do the changes in #480 do anything to improve the situation?

**Comment by whhhooop** [comment]
Yes it helps. The result type is
```
(state: any) => {
    foo: string;
    bar: number;
}
```
It doesn't infer the type of `state` but i'm not sure if it's supposed to.

**Comment by whhhooop** [comment]
No. Although the error message has changed.
```
Type '(state: GlobalState) => string' is not assignable to type '(state: unknown) => string'.
  Types of parameters 'state' and 'state' are incompatible.
    Type '{}' is missing the following properties from type 'GlobalState': foo, bar
```

**Comment by eXamadeus** [comment]
So it looks like the original typing of `createSturcturedSelector` was meant to take in type params to correctly generate the output.

You basically want this:

```typescript
import { createStructuredSelector } from 'reselect';

type GlobalState = {
  foo: string;
  bar: number;
};

const selectFoo = (state: GlobalState) => state.foo;
const selectBar = (state: GlobalState) => state.bar;

// <Input, Output>
const injectState = createStructuredSelector<GlobalState, { foo: string, bar: number }>({
  foo: selectFoo, // error
  bar: selectBar, // error
});
```

where `Input` is `GlobalState` and `Output` is whatever the structured selector will map too (in this case, it's the same as `GlobalState` but it could be different).

Basically for this typing to work, you need to provide the "input" and "output" types. @markerikson we should note this as something we might be able to improve in #484. Not sure how easy it will be though.

**Comment by eXamadeus** [comment]
@markerikson I'll add a test for this in #486 like this:

```typescript
function testStructuredSelectorTypeParams() {
  type GlobalState = {
    foo: string;
    bar: number;
  };

  const selectFoo = (state: GlobalState) => state.foo;
  const selectBar = (state: GlobalState) => state.bar;

  // Output state should be the same as input, if not provided
  // @ts-expect-error
  createStructuredSelector<GlobalState>({
    foo: selectFoo,
    // bar: selectBar,
    // ^^^ because this is missing, an error is thrown
  });
}
```

**Comment by markerikson** [comment]
_Should_ be fixed by #486 .
